### DISCUSSION TITLE
Lastmod support

### DISCUSSION THREAD

**Discussion by mkbctrl** [ReeceM/prismic-sitemap#8] [Q&A]
Recently I was updating few sitemaps in order to meet updated google's requirements, where the `changefreq` and `priority` fields are kind of obsolete. 

First I stumbled upon this information while working with another project in GatsbyJs:
![Screenshot 2021-06-24 at 17 50 27](https://user-images.githubusercontent.com/10107362/123293985-ba80d680-d514-11eb-97fe-652ecf62f5c1.png)

source: https://www.gatsbyjs.com/plugins/gatsby-plugin-sitemap/
direct source: https://developers.google.com/search/docs/advanced/sitemaps/build-sitemap

Since I used your plugin today, I thought I will share this info with you, as I don't see this field in the output file. Hope you find it useful.

Btw, it's a great plugin, and thanks for saving me a lot of work ❤️ 

P.S. I am about to have a look how to add the `lastmod` field into your sitemap, hopefully I will be able to share some more insights.

**Comment by ReeceM** [comment]
Hi @mkbctrl,

Thank you for highlighting this. I am not aware of that and I think I will definitely make it an addition then, at least with the Prismic CMS data there is an updated date field that can be passed in. 

At the moment, there is the option to use the `optionsMapPerDocumentType` object, but this doesn't have access to the current document data that is being passed in. 

I am thinking then that I could make that then a function that can have the current document passed in, but also still be backwards compatible with the object only version.

So something like this: 

```javascript
// next.config.js


module.exports = withPrismicSitemap({
  sitemap: {
    // ... other cofing
    optionsMapPerDocumentType: {
      post: (doc) => {
        return {
          lastmod: doc.last_publication_date ? doc.last_publication_date : (new Date()).toJSON(),
          // below for any older things that use it
          changefreq: "weekly",
          priority: 0.8
        };
      }
    },
  }
})

```

I will have to have a look at this then as a very useful and practical feature, thank you for highlighting it 👍

**Reply by mkbctrl** [reply]
Your idea looks sharp, I think it's exactly what's required here. Since we are "speaking", in my project not all the pages are coming from prismic, some of them live only inside the nextjs app. Are you aware of a workaround to append such pages to your `sitemap.xml`? Even looping through a manually created array would be good enough, I think.

**Reply by ReeceM** [reply]
Sorry for the delay a between responses. At the moment, I am not aware of a work around with the current version, but I am having a look at it now though. 👍 

What you have highlighted with the usage of some next static pages that exist already was something I was wondering about, but hadn't implemented yet. 

Are those static pages, or dynamic ones? I am just wondering if allowing passing a extra list of routes for other pages or appending to the sitemap before writing with another function that can be called.

**Comment by ReeceM** [comment]
Hi @mkbctrl,

I have added the option to allow a callback to be placed inside the options map. There is also the addition to allow someone to place the usage of static paths in an array for them to be added to the list of the sitemap entries.

If you would like to test it, it will be available as `@reecem/prismic-sitemap@^0.3.0-alpha.1`.

If you would like you can test it and let me know if it is coming out alright for you too. You can see PR #10 for the changes that have been added and also the branch

**Reply by ReeceM** [reply]
https://www.npmjs.com/package/@reecem/prismic-sitemap/v/0.3.0-alpha.1

**Reply by mkbctrl** [reply]
Thanks a lot! I have booked the time for this tomorrow. I will let you know then.

**Reply by mkbctrl** [reply]
I would change the date format to YYYY-MM-DD, here's a sample from Google docs:

```
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>http://www.example.com/foo.html</loc>
    <lastmod>2018-06-04</lastmod>
  </url>
</urlset>
```

I am not sure if it's a real issue, but you can find claims that misrepresented date will be ignored... so I leave it to you, to decide :)

**Reply by mkbctrl** [reply]
With static paths I would allow to pass a function as parameter as well. Not that you can't do it, but later on you try to format it on your own with: `smi.lastmod = new Date(smiLoose.lastmod).toISOString();`

Why? Developer might want to pass `Date.now()` or similar to have a fresh date on every deploy.

So you can either solve it by yourself or give the flexibility to end user.

**Reply by mkbctrl** [reply]
Other than that, everything works smoothly. Thanks a lot for great support 🍻 🍻 🍻 🍻 🍻 🍻 🍻

**Reply by ReeceM** [reply]
Hi @mkbctrl, thank you for the feedback.

I will have a look at the usage of the date format in `YYYY-MM-DD`, I might have to suggest that in the docs, or pass a helper function to convert the date to that format as the an end user would be returning that result, will have a look.

Regarding the static paths, I think the addition of allowing handling functions including the object might be better, I was wondering about adding that. I think it might be more consistent then.

Although in the current state, if someone would like the date updated each deploy for a static path, they can set the `next.config.js` file with the following:

```javascript
{
    // other stuff
    staticPaths: [
        {
            url: 'example-1',
            lastmod: Date.now(),
        },
        {
            url: 'example-2',
            lastmod: '2021-01-01',
        },
        {
            url: 'example-3',
            lastmod: (new Date).toISOString(),
        }
    ]
}
```

But will have a look at passing a function and may make it async/await capable too if there is a page queried from elsewhere then.

Thank you again for the feedback and highlighting this 👍

**Reply by ReeceM** [reply]
The other option is to enable: 
```javascript

// next.config.js

module.exports = withPrismicSitemap({
  sitemap: {

    sitemapConfig: {
      lastmodDateOnly: true,
    },

  }
})

```

This affects the underlying library that builds the sitemap stream. This is also an option as it will force all dates to be the same and then there isn't a mix of the date formats.

**Reply by mkbctrl** [reply]
> The other option is to enable:
> 
> ```js
> // next.config.js
> 
> module.exports = withPrismicSitemap({
>   sitemap: {
> 
>     sitemapConfig: {
>       lastmodDateOnly: true,
>     },
> 
>   }
> })
> ```
> 
> This affects the underlying library that builds the sitemap stream. This is also an option as it will force all dates to be the same and then there isn't a mix of the date formats.

And I think that's how you should tackle this

**Reply by ReeceM** [reply]
> And I think that's how you should tackle this

I will make it enabled by default if someone would want to disable it for some reason if they want the time as well. 👍 💯

**Comment by mkbctrl** [comment]
One not related issue, while trying to install your package in alpha version I was forced to switch to node v12, by default I run LTS v14. If it's not required by dependencies, maybe it's worth to change.

**Reply by mkbctrl** [reply]
My linter on gitlab CI doesn't like it as well:
```
error @reecem/prismic-sitemap@0.3.0-alpha.1: The engine "node" is incompatible with this module. Expected version "^12.0". Got "14.1.0"
error Found incompatible module.
```

**Reply by ReeceM** [reply]
That would be my fault, I think I misconfigured the node version stuff in wanting to make sure someone didn't have issues trying to use it on v10 of Node, will fix it now though, thank you 👍

**Reply by mkbctrl** [reply]
thanks 🍻 , please let me know once you find the time

**Reply by ReeceM** [reply]
will do 👍

**Reply by ReeceM** [reply]
version 0.3.0-alpha.2 should have the fixed things for the nom engine and the lastmod.

Works on my v14 node install

**Comment by ReeceM** [comment]
Hi @mkbctrl

Hope you are good, I released v0.3.0 yesterday with these additions and fixes, so if you would like that is the version that will get updates for any fixes now. (I won't push an alpha for that version now)

Hope it works and please let me know if there are any issues or feel free to open an issue if you would like to put more details in 👍

**Reply by mkbctrl** [reply]
I just installed the latest version (0.3) and took it for a ride. So far so good. Everything works as expected. I would like to thank you for your great support once again :))

**Reply by ReeceM** [reply]
Awesome 💯. I am glad it's working good then 👍
